myapp/utility.py

```diff
@@ -1,116 +1,880 @@
+
+
+def experiment_exists(experiment_id):
+    return ds.query(m.Experiment).filter(m.Experiment.id == experiment_id).count() > 0
+
+
+def observation_exists(observation_id):
+    return ds.query(m.Observation).filter(m.Observation.id == observation_id).count() > 0
+
+
+def class_exists(class_id):
+    return ds.query(m.Class).filter(m.Class.id == class_id).count() > 0
+
+
+def user_exists(user_id):
+    return ds.query(m.User).filter(m.User.id == user_id).count() > 0
+
+
+# --- Query data ---
+# Get the Lab object from the database by lab_id
+def query_lab(lab_id):
+    return ds.query(m.Lab).filter(m.Lab.id == lab_id).first()
+
+
+def query_experiment(experiment_id):
+    return ds.query(m.Experiment).filter(m.Experiment.id == experiment_id).first()
+
+
+# Get the Observation object from the database by observation_id
+def query_observation(observation_id):
+    return ds.query(m.Observation).filter(m.Observation.id == observation_id).first()
+
+
+def query_user(user_id):
+    return ds.query(m.User).filter(
+        m.User.id == user_id).first()
+
+
+def query_class(class_id):
+    return ds.query(m.Class).filter(m.Class.id == class_id).first()
+
+
+def query_role(role_name):
+    return ds.query(m.Role).filter(
+        m.Role.name == role_name).first()
+
+
+def query_all_lab():
+    return ds.query(m.Lab).all()
+
+
+def query_all_experiment():
+    return ds.query(m.Experiment).all()
+
+
+# return the query for all the classes in the database
+def query_all_class():
+    return ds.query(m.Class).all()
+
+
+# return the query for all the admins in the database
+def query_all_admin():
+    return ds.query(m.User).filter(
+           m.User.role_name == 'Admin').all()
+
+
+# return the query for all the admins in the database
+def query_all_superadmin():
+    return ds.query(m.User).filter(
+           m.User.role_name == 'SuperAdmin').all()
+
+
+# Get lab query for the user according to user's role
+# A SuperAdmin can select among all the labs
+    # A student/Admin can only select his/her own labs
+def query_labs_for_user(user):
+    if user.role_name == 'SuperAdmin':
+        return ds.query(m.Lab).all()
+    else:
+        return [lab for lab in user.labs]
+
+
+# Get all the experiments for a lab
+def query_experiments_for_lab(lab_id):
+    return ds.query(m.Experiment).filter(
+        m.Experiment.lab_id == lab_id).order_by(m.Experiment.order).all()
+
+
+# Get all the observations for an experiment
+def query_observations_for_experiment(experiment_id):
+    return ds.query(m.Observation).filter(
+        m.Observation.experiment_id == experiment_id).order_by(
+        m.Observation.id).all()
+
+
+# --- Check number ---
+def observation_number_for_experiment(experiment_id):
+    return db.session.query(m.Observation).filter(
+        m.Observation.experiment_id == experiment_id).count()
+
+
+# ---serialization/deserialization---
+# convert a query for labs into a list of JSON objects
+def serialize_lab_list(lab_list_query):
```

Comment:
this function doesn't do what you say it does-- it's hydrating lab_list_query into a python list of dictionaries.